### PR TITLE
Add lightgbm package

### PR DESCRIPTION
**Pull request by yzhliu**
### Description

add the lightgbm package. related issue #3116 .

Question though: I used patch to specify setup.py argument (--noomp), wondering if there's a better way to do so?

### Checklists

- [x] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry
- [x] Add / update tests

### CODE REVIEW

**Comment by hoodmane** [comment]
I think there's a key called `backend-args` where you should put `--noomp`.

File: packages/lightgbm/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package:
+  name: lightgbm
+  version: 3.3.2
+source:
+  url: https://files.pythonhosted.org/packages/8f/51/d272fa065be3be615d279be915705fa3824b86155e36c974cfb8d3ceec1e/lightgbm-3.3.2.tar.gz
+  sha256: 5d25d16e77c844c297ece2044df57651139bc3c8ad8c4108916374267ac68b64
+  patches:
+    - patches/0001-noomp-True.patch
+build:
+  script: |
```
Comment (hoodmane) [inline packages/lightgbm/meta.yaml:10]:
```suggestion
  backend-flags: --noomp
  script: |
```

File: packages/lightgbm/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    export CMAKE_TOOLCHAIN_FILE=$PYODIDE_ROOT/packages/lightgbm/cmake/Toolchain.cmake
+  exports: "whole_archive"
```
Comment (hoodmane) [inline packages/lightgbm/meta.yaml:12]: Does it not work with `exports: "requested"`?

**Comment by hoodmane** [comment]
Also, maybe this can be simplified after we merge #3128 @ryanking13?

**Comment by ryanking13** [comment]
> Also, maybe this can be simplified after we merge #3128 @ryanking13?

Yes, I think we won't need a toolchain file after #3128.

**Comment by ryanking13** [comment]
Thanks for the porting work @yzhliu. We merged #3128 so I think you won't need to use a custom CMAKE_TOOLCHAIN_FILE now.

**Comment by yzhliu** [comment]
@ryanking13 @hoodmane cool, I'll change accordingly.

File: packages/lightgbm/meta.yaml (hunk 2 above)
Comment (yzhliu) [reply packages/lightgbm/meta.yaml:12]: It'll fail saying "OSError: [Errno 7] Argument list too long: 'em++'"

**Comment by yzhliu** [comment]
@hoodmane @ryanking13 could you do another round of code review? and how shall I fix the test coverage check?

File: pyodide-build/pyodide_build/pypabuild.py [hunk 3]
```diff
@@ -98,8 +98,17 @@ def _build_in_isolated_env(
         symlink_unisolated_packages(env)
         install_reqs(env, builder.build_system_requires)
         installed_requires_for_build = False
+        # "--build-option" contains custom options that only apply to current package,
+        # passing it to the dependencies may cause error.
+        # e.g., lightgbm accepts "--nomp" as one of its setup.py arguments,
+        # but "--nomp" isn't supported by its dependencies such as numpy, scipy, etc.
+        reqs_config_settings = {
+            k: v for k, v in config_settings.items() if k != "--build-option"
```
Comment (ryanking13) [inline pyodide-build/pyodide_build/pypabuild.py:106]: Is our `backend-flags` propagating to dependencies? That seems very weird.

Comment (ryanking13) [reply pyodide-build/pyodide_build/pypabuild.py:106]: Ah, okay we are passing backend_flags in `builder.get_requires_for_build`. I think we should only pass it to `builder.build` and not  `builder.get_requires_for_build` @hoodmane?

**Comment by ryanking13** [comment]
Thanks @yzhliu! I think the test coverage is okay. I think the backend flags passed to the dependencies is a bug and otherwise it LGTM.

File: pyodide-build/pyodide_build/pypabuild.py (hunk 3 above)
Comment (hoodmane) [reply pyodide-build/pyodide_build/pypabuild.py:106]: I certainly didn't mean for it to propagate to dependencies!

Comment (ryanking13) [reply pyodide-build/pyodide_build/pypabuild.py:106]: > I certainly didn't mean for it to propagate to dependencies!

Okay, thanks for the confirmation. @yzhliu Could you remove all backend-flags(config_settings) passed to `builder.get_requires_for_build`? Thanks for catching this out!

Comment (yzhliu) [reply pyodide-build/pyodide_build/pypabuild.py:106]: sure, I'll do a sweep over the related code.

Comment (yzhliu) [reply pyodide-build/pyodide_build/pypabuild.py:106]: @ryanking13 seems here are the only two usage of `builder.get_requires_for_build` in the code base.

Comment (ryanking13) [reply pyodide-build/pyodide_build/pypabuild.py:106]: Yes, there are two appearances of `builder.get_requires_for_build`.

Comment (yzhliu) [reply pyodide-build/pyodide_build/pypabuild.py:106]: I changed both of them, is the PR good to go? or did I miss anything else?

Comment (ryanking13) [reply pyodide-build/pyodide_build/pypabuild.py:106]: Is it pushed? I can't notice the change and it looks like the last push you made is 3 days ago before the above comments.

Comment (yzhliu) [reply pyodide-build/pyodide_build/pypabuild.py:106]: no the original commit already has the two changes,
https://github.com/pyodide/pyodide/pull/3138/files#diff-8a5df82b2fccc945f9e52b43277a75fa9e7a3f3ea1f5b03725b97d186f84225fR110
https://github.com/pyodide/pyodide/pull/3138/files#diff-8a5df82b2fccc945f9e52b43277a75fa9e7a3f3ea1f5b03725b97d186f84225fR122

**Review by ryanking13 — APPROVED** [review]
Thanks again @yzhliu!

**Pull request merged**